### PR TITLE
health check: adding support for basic mysql health checks

### LINKED ISSUE #3501: Passwordless MySQL Health Check
*Title*: Passwordless MySQL Health check

*Description*:
For feature parity with HAProxy, Envoy should have support for basic mysql health check. It involves sending a login request for a passwordless user, followed by a quit request.

*Relevant Links*:
- HAProxy documentation: http://cbonte.github.io/haproxy-dconv/1.5/configuration.html#4-option%20mysql-check
- User group discussion: https://groups.google.com/forum/#!searchin/envoy-users/feng$20pan%7Csort:date/envoy-users/SVjtIHa--Lo/YmdpaHvNAQAJ

### PR DESCRIPTION
**Pull request by juchem**
*Description*:
Adding support for basic mysql health checks, for feature parity with
HAProxy. Only passowrdless users are supported, so it's generally a good
idea to create a db user with no privileges exclusively for the health
checks.

*Risk Level*: Medium

*Testing*:
Unit tests added for the following scenarios:
{reuse_connection, dont_reuse_connection} x ({success, timeout, remote_close}  + { bad_server_greeting, bad_server_reply} x coverage)

*Docs Changes*:
Inline documentation added for the new health check API.

*Release Notes*:
Added an entry to `version_history.rst` with the new feature.

cc: @lap1817

Fixes #3501

### CODE REVIEW

**Comment by mattklein123** [comment]
@juchem thanks for the contribution. This new health checker should be implemented with our new health checker extension system. See the Redis one for example: https://github.com/envoyproxy/envoy/tree/master/source/extensions/health_checkers

Can you convert it over to an optional extension in a new mysql directory? Thank you!

**Comment by juchem** [comment]
sure, working on it

**Comment by juchem** [comment]
@mattklein123 done

**Comment by mattklein123** [comment]
@juchem sorry for the delay here. I will review by the end of the week.

**Comment by mattklein123** [comment]
@rshriram I'm going to take a pass on this now. Is anyone from VMWare interested in helping to review?

**Comment by mattklein123** [comment]
Tagging @rshriram as I think he will help review also.

File: api/envoy/config/health_checker/mysql/v2/mysql.proto [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+package envoy.config.health_checker.mysql.v2;
+option go_package = "v2";
+
+// [#protodoc-title: MySQL]
+// MySQL health checker :ref:`configuration overview <config_health_checkers_mysql>`.
+
+// Configuration for MySQL health check. The health check involves authenticating against the MySQL
+// server with a passwordless user and closing the session immediatelly after, while validating the
+// server response. It's recommended to create an user with no privileges, exclusively for the
```
Comment (mattklein123) [inline api/envoy/config/health_checker/mysql/v2/mysql.proto]: s/an user/a user

File: api/envoy/config/health_checker/mysql/v2/mysql.proto [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+// health check operation.
+message MySQL {
+  // Username to authenticate with for health checking
```
Comment (mattklein123) [inline api/envoy/config/health_checker/mysql/v2/mysql.proto]: nit: period end of sentence.

File: api/envoy/config/health_checker/mysql/v2/mysql.proto [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  string user = 1;
```
Comment (mattklein123) [inline api/envoy/config/health_checker/mysql/v2/mysql.proto]: Can we add a validation that this is not empty?

File: docs/root/configuration/health_checkers/mysql.rst [hunk 4]
```diff
@@ -0,0 +1,13 @@
+.. _config_health_checkers_mysql:
+
+MySQL
+=====
+
+The MySQL health checker is a custom health checker which checks MySQL upstream hosts. It sends
+a MySQL LOGIN command followed by a QUIT command, so that the session is ended properly, and expects
+an OK response. The upstream MySQL server can respond with anything other than OK to cause an
+immediate active health check failure. An user name must be provided for the LOGIN command to
```
Comment (mattklein123) [inline docs/root/configuration/health_checkers/mysql.rst]: a/An user name/A user name

File: source/common/common/byte_order.h [hunk 5]
```diff
@@ -6,13 +6,70 @@
 
 #include <libkern/OSByteOrder.h>
 
+#define htole16(x) OSSwapHostToLittleInt16((x))
```
Comment (mattklein123) [inline source/common/common/byte_order.h]: Can you merge master also when you update? I think @zuercher just added these in his Thrift PR.

File: source/common/common/byte_order.h [hunk 6]
```diff
@@ -6,13 +6,70 @@
...
 
 #include <endian.h>
 
 #endif
+
+#include <climits>
+
+template <typename T, size_t = sizeof(T) * CHAR_BIT> struct endianness_converter;
+
+// convenience function that converts a host byte-order integral to little endian byte-order
```
Comment (mattklein123) [inline source/common/common/byte_order.h]: nice!

File: api/envoy/config/health_checker/mysql/v2/BUILD [hunk 7]
```diff
@@ -0,0 +1,8 @@
+load("//bazel:api_build_system.bzl", "api_proto_library")
```
Comment (mattklein123) [inline api/envoy/config/health_checker/mysql/v2/BUILD]: Check out the recent thrift PR. Can we add this extension to CODEOWNERS?

File: source/extensions/health_checkers/mysql/mysql.cc [hunk 8]
```diff
@@ -0,0 +1,323 @@
+#include "extensions/health_checkers/mysql/mysql.h"
```
Comment (mattklein123) [inline source/extensions/health_checkers/mysql/mysql.cc]: There is some missing coverage for error cases in this file. Can you check the coverage build in CI and add some more tests?

File: source/extensions/health_checkers/mysql/mysql.h [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    ~MySQLActiveHealthCheckSession();
+
+    void onData(Buffer::Instance& data);
+    void onEvent(Network::ConnectionEvent event);
+
+    // ActiveHealthCheckSession
+    void onInterval() override;
+    void onTimeout() override;
+
+    bool parse_mysql_server_greeting_packet(Buffer::Instance& data);
```
Comment (mattklein123) [inline source/extensions/health_checkers/mysql/mysql.h]: nit: can we make all these methods camelCase per normal Envoy style?

File: source/extensions/health_checkers/mysql/mysql.cc [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+
+  // we can't assume unaligned reads are permitted so we trust the standard library to efficiently
+  // perform the copy for small-sized buffers
+  std::copy(data, data + Bytes, reinterpret_cast<uint8_t*>(&result));
+
+  return from_little_endian(result);
+}
+
+template <size_t Bytes, typename T>
+static inline void write_little_endian_int(Buffer::Instance& out, T value) {
```
Comment (mattklein123) [inline source/extensions/health_checkers/mysql/mysql.cc]: I don't know the answer to this, but between the Thrift, Mongo, and now this, is there any way to share any of this type of functionality in a common EndianBufferHelper or something? cc @zuercher

File: source/extensions/health_checkers/mysql/mysql.cc [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+bool MySQLHealthChecker::MySQLActiveHealthCheckSession::parse_mysql_ok_packet(
+    Buffer::Instance& data) {
+  // parse response packet
+  // https://dev.mysql.com/doc/internals/en/packet-OK_Packet.html
+  constexpr const auto MYSQL_OK_RESPONSE = 0;
+  constexpr const size_t ok_response_size = 1;
+  constexpr const size_t ok_packet_size_threshold = 7;
+
+  const size_t minimum_needed_packet_size = mysql_packet_header_size + ok_response_size;
+  if (data.length() < minimum_needed_packet_size) {
```
Comment (mattklein123) [inline source/extensions/health_checkers/mysql/mysql.cc]: Technically, I think you need to handle partial data here (and elsewhere in similar parsing functions)? There is no guarantee that all the data will come in 1 shot. If that is something that you don't think we need to handle, we should clearly document that and possibly add a TODO here, though I would rather we just handle it. Thoughts?

File: source/extensions/health_checkers/mysql/mysql.cc (hunk 10 above)
Comment (zuercher) [reply source/extensions/health_checkers/mysql/mysql.cc]: I think if we can agree on parsing directly from the buffer (a la mongo/thrift) vs linearizing the entire request before parsing (mysql), we could come up with a helper that supports all 3 protocols. Thrift requests can be almost arbitrarily large, and depending on configuration the size may not be known at the start of the request, so it's not reasonable to linearize them.

Between the three I already see 8/16/24/32/64-bit integers (some signed, some unsigned) plus both null- and non-terminated strings, so there's a lot to go into this utility already.

Comment (mattklein123) [reply source/extensions/health_checkers/mysql/mysql.cc]: Yeah, +1. @juchem can we at least file a tech debt tracking issue on figuring this out?

**Comment by juchem** [comment]
Hey guys, thanks for all the feedback. I just got out on paternal leave so it'll take me a while until I can work on this again.

File: source/extensions/health_checkers/mysql/mysql.cc [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+
+  // If NUL not found or at first position in the payload (protocol version 0, which doesn't
+  // exist) the packet is malformed.
+  if (server_version_nul <= static_cast<ssize_t>(mysql_packet_header_size)) {
+    ENVOY_CONN_LOG(trace, "mysql healthcheck failed [greeting]: malformed packet ", *client_);
+    return false;
+  }
+
+  // We calculate the amount of data expected to be in the packet and do a best-effort to
+  // linearize as much as possible in one-shot, but no more.
```
Comment (rshriram) [inline source/extensions/health_checkers/mysql/mysql.cc]: What does linearize mean here? Deserialize the payload ? If so I would rephrase accordingly. Because a stream of bytes is inherently linear.

File: source/extensions/health_checkers/mysql/mysql.cc [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+    data.drain(data.length());
+    return;
+  }
+
+  phase_ = Phase::Over;
+  data.drain(data.length());
+
+  handleFailure(FailureType::Active);
+  client_->close(Network::ConnectionCloseType::NoFlush);
+}
```
Comment (rshriram) [inline source/extensions/health_checkers/mysql/mysql.cc]: A whole lot of functions here can be useful for anyone wanting to build a full blown MySQL codec in Envoy. I suggest moving these useful MySQL parsing stuff (login, packet okay and others) into a generic MySQL folder (I don’t know the right destination for it given that there is no MySQL filter), but it would provide the filter implementor a jump start. With all the codec functions abstracted, you could encapsulate checks for partial packets etc into the codec, such that the health check is dramatically simplified into 
 MySQL.doLogin(user)
 MySQL.resp == okay -> health check ok

**Comment by mattklein123** [comment]
I think @juchem is on paternity leave. I will just close for now and we can reopen when he is back.

File: source/extensions/health_checkers/mysql/mysql.cc (hunk 13 above)
Comment (juchem) [reply source/extensions/health_checkers/mysql/mysql.cc]: sounds like a good idea. I'm not sure that would be useful out-of-the-box for a filter implementor, but it's a starting skeleton as you mentioned. I'll see what I can come up with

File: source/extensions/health_checkers/mysql/mysql.cc (hunk 12 above)
Comment (juchem) [reply source/extensions/health_checkers/mysql/mysql.cc]: linearize is a Buffer operation. I'll check Mongo's implementation as mentioned somewhere else to see how it reads straight from the buffer

File: source/extensions/health_checkers/mysql/mysql.cc (hunk 11 above)
Comment (juchem) [reply source/extensions/health_checkers/mysql/mysql.cc]: I haven't seen the need for it just yet, so I believe leaving it as a TODO might suffice. As for how partial data is handled elsewhere in Envoy, does it accumulate partial data in a local buffer and parses in one shot or is there support for a parsing state machine? Implementing such a machine by hand sounds counter-productive and error prone.

File: source/extensions/health_checkers/mysql/mysql.cc (hunk 10 above)
Comment (juchem) [reply source/extensions/health_checkers/mysql/mysql.cc]: I agree it's better. I'll check Mongo/Thrift implementation to see how they do it and depending on the work I'll either file a tech debt or update the PR.

File: source/extensions/health_checkers/mysql/mysql.h (hunk 9 above)
Comment (juchem) [reply source/extensions/health_checkers/mysql/mysql.h]: of course. sorry about that, it's just muscle memory

File: source/extensions/health_checkers/mysql/mysql.cc (hunk 8 above)
Comment (juchem) [reply source/extensions/health_checkers/mysql/mysql.cc]: will do

File: api/envoy/config/health_checker/mysql/v2/BUILD (hunk 7 above)
Comment (juchem) [reply api/envoy/config/health_checker/mysql/v2/BUILD]: that might have been added by mistake while rebasing

File: source/common/common/byte_order.h (hunk 5 above)
Comment (juchem) [reply source/common/common/byte_order.h]: will do

File: docs/root/configuration/health_checkers/mysql.rst (hunk 4 above)
Comment (juchem) [reply docs/root/configuration/health_checkers/mysql.rst]: good catch

File: api/envoy/config/health_checker/mysql/v2/mysql.proto (hunk 3 above)
Comment (juchem) [reply api/envoy/config/health_checker/mysql/v2/mysql.proto]: I believe an empty string is actually a valid case for an anonymous user auth

File: api/envoy/config/health_checker/mysql/v2/BUILD (hunk 7 above)
Comment (juchem) [reply api/envoy/config/health_checker/mysql/v2/BUILD]: nvm, I've mixed things up in my head

Comment (mattklein123) [reply api/envoy/config/health_checker/mysql/v2/BUILD]: https://github.com/envoyproxy/envoy/blob/master/CODEOWNERS. See the new extension policy. Each extension needs owners that will handle code reviews.

Comment (juchem) [reply api/envoy/config/health_checker/mysql/v2/BUILD]: I can add it to that file. Who should I put up as owner?

Comment (juchem) [reply api/envoy/config/health_checker/mysql/v2/BUILD]: :)
kinda saw that coming, but as I transition to a new team I won't have the cycles to spare, unfortunately

Comment (mattklein123) [reply api/envoy/config/health_checker/mysql/v2/BUILD]: @juchem is there someone on your old team who would want to help maintain this? If not, another option is to just keep it as a private or public out of tree filter for now.

Comment (juchem) [reply api/envoy/config/health_checker/mysql/v2/BUILD]: @austin-zhu @Jason-Jian @ken-experiment any interest in helping maintain the MySQL health check?

Comment (Jason-Jian) [reply api/envoy/config/health_checker/mysql/v2/BUILD]: Sign me up. although I'd need to come up to speed first. :)

Comment (juchem) [reply api/envoy/config/health_checker/mysql/v2/BUILD]: There you go. I'll spend some time this week with Jason to provide an overview of the code structure, then I'll add him as a maintainer of the MySQL health checker in the `CODEOWNERS` file.

File: source/extensions/health_checkers/mysql/mysql.cc (hunk 8 above)
Comment (juchem) [reply source/extensions/health_checkers/mysql/mysql.cc]: @mattklein123 where can I find the coverage build in CI? I can't find the build checks. Maybe because the PR was closed?

**Comment by mattklein123** [comment]
@juchem I reopened it. If you push new commits it should build again.

File: source/extensions/filters/network/thrift_proxy/buffer_helper.h [hunk 14]
```diff
@@ -69,104 +69,104 @@ class BufferHelper {
    * @param offset offset into buffer to peek at
    * @return the int8_t at offset in buffer
    */
-  static int8_t peekI8(Buffer::Instance& buffer, uint64_t offset = 0);
+  static int8_t peekBEI8(Buffer::Instance& buffer, uint64_t offset = 0);
```
Comment (zuercher) [inline source/extensions/filters/network/thrift_proxy/buffer_helper.h]: I don't think the big- or little-endian appellation applies to single bytes, so probably undo this one.

But for the rest of the methods, is there a specific reason for renaming them all? For example, are you planning on moving this helper class out of the thrift extension so it can be reused elsewhere?

**Comment by juchem** [comment]
It is a preparation to move it outside of thrift. Renaming the 8 bit
version was mostly due to uniformity, but I agree it might have been a bad
idea.
Btw, I forgot to tag the PR as WIP since I still have more changes to make.
How can I cause the coverage tests to kick in?

**Comment by zuercher** [comment]
You have to add a commit to get CI to run again, unfortunately.

**Comment by juchem** [comment]
@mattklein123 I'm confused about how to look at the coverage test output. There's a rather large output file but all I see are unit test results as OK for MySQL. I'm probably missing something obvious, but I didn't find any coverage information there.

**Comment by dio** [comment]
@juchem sorry for jumping in, but I think this one: https://73240-65214191-gh.circle-artifacts.com/0/build/envoy/generated/coverage/coverage.source_extensions_health_checkers_mysql_mysql.cc.html needs more loves in terms of coverage.

**Comment by dio** [comment]
One way to browse the coverage is by clicking the "Artifacts" tab. E.g. https://circleci.com/gh/envoyproxy/envoy/73240#artifacts/containers/0

image

**Comment by juchem** [comment]
I don't seem to have access to this tab for some reason.
image

**Comment by dio** [comment]
@juchem you could try to log-in with Github to Circle-CI.

**Comment by juchem** [comment]
I believe I am
image

**Comment by juchem** [comment]
Ah, that was the problem: I was logged with my corporate account, not the personal one.

**Comment by juchem** [comment]
took care of the coverage, it was a simple issue in the way test packets were built

**Comment by mattklein123** [comment]
@juchem so is this ready for re-review?

**Comment by juchem** [comment]
Yes

**Comment by mattklein123** [comment]
@juchem can you merge master? Then I promise I will look at this PR this week. Thank you!

**Comment by juchem** [comment]
alright. I'll rebase and make sure tests run, then @Jason-Jian will commandeer the PR

**Comment by mattklein123** [comment]
Sorry will look at this soon.

File: docs/root/intro/version_history.rst [hunk 15]
```diff
@@ -101,6 +101,7 @@ Version history
   <envoy_api_field_core.HealthCheck.unhealthy_edge_interval>`, :ref:`unhealthy to healthy
   <envoy_api_field_core.HealthCheck.healthy_edge_interval>` and for subsequent checks on
   :ref:`unhealthy hosts <envoy_api_field_core.HealthCheck.unhealthy_interval>`.
+* health check: added basic support for :ref:`MySQL health checks <config_health_checkers_mysql>`.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: Should be moved to the 1.8.0 section.

File: include/envoy/buffer/buffer.h [hunk 16]
```diff
@@ -179,6 +182,118 @@ class Instance {
...
    */
   virtual Api::SysCallResult write(int fd) PURE;
+
+  /**
+   * Copy an integral out of the buffer.
+   * @param start supplies the buffer index to start copying from.
+   * @param Size how many bytes to read out of the buffer.
+   * @param Endianness specifies the byte order to use when decoding the integral.
+   */
+  template <typename T, ByteOrder Endianness = ByteOrder::Host, size_t Size = sizeof(T)>
```
Comment (mattklein123) [inline include/envoy/buffer/buffer.h]: Nice. I really like this set of changes. Can we please move this change (and the associated Thrift changes) into a separate change? We should make sure the new buffer functions have explicit tests in buffer_impl_test. Let's get that change first and then we can merge master into this change and get this in. Thank you!

**Comment by Jason-Jian** [comment]
@mattklein123 FYI there were conflicts with upstream when I tried to rebase, I haven't fully resolved them yet. To clarify, your suggestion was to separate this PR into 2, with explicit tests for the buffer functions, correct? I can reopen the PR once I get these resolved.

**Comment by zuercher** [comment]
@Jason-Jian Matt's out for a bit, but I think I understand what he wanted:

PR 1: the include/envoy/buffer and source/common/byte_order changes, and the refactoring of thrift_proxybuffer_helper.h/cc to use them. This PR should include direct tests for the byte_order code in test/common (it may be that you can move test cases from buffer_helper_test.cc, it's up to you)

PR 2: the mysql health check extension and it's API changes.

The idea is that as much as possible, PRs related to extensions should avoid modifying main-line Envoy code.
